### PR TITLE
Fix error message for mismatched types

### PR DESCRIPTION
**Pull request by alexeyzab**
This addresses #41425 by implementing the changes mentioned in the
following comment:
https://github.com/rust-lang/rust/issues/41425#issuecomment-296754508

### CODE REVIEW

File: src/librustc_typeck/check/coercion.rs [hunk 1]
```diff
@@ -1119,11 +1127,7 @@ impl<'gcx, 'tcx, 'exprs, E> CoerceMany<'gcx, 'tcx, 'exprs, E>
                 }
             }
             Err(err) => {
-                let (expected, found) = if expression.is_none() {
-                    // In the case where this is a "forced unit", like
-                    // `break`, we want to call the `()` "expected"
-                    // since it is implied by the syntax.
-                    assert!(expression_ty.is_nil());
+                let (expected, found) = if label_expression_as_expected {
                     (expression_ty, self.final_ty.unwrap_or(self.expected_ty))
```
Comment (arielb1) [inline src/librustc_typeck/check/coercion.rs]: nit: you dropped half of a comment

File: src/test/ui/coercion-missing-tail-expected-type.stderr [hunk 2]
```diff
@@ -0,0 +1,19 @@
+error[E0308]: mismatched types
+  --> $DIR/coercion-missing-tail-expected-type.rs:13:28
+   |
+13 |   fn plus_one(x: i32) -> i32 {
+   |  ____________________________^
+14 | |     x + 1;
+15 | | }
+   | |_^ expected (), found i32
+   |
+   = note: expected type `i32`
```
Comment (arielb1) [inline src/test/ui/coercion-missing-tail-expected-type.stderr]: why are the note and label the other way around?

File: src/librustc_typeck/check/coercion.rs (hunk 1 above)
Comment (alexeyzab) [inline src/librustc_typeck/check/coercion.rs]: This part of the comment was meant to be deleted per https://github.com/rust-lang/rust/issues/41425#issuecomment-296754508. Should I rephrase the remaining part?

File: src/test/ui/coercion-missing-tail-expected-type.stderr (hunk 2 above)
Comment (alexeyzab) [inline src/test/ui/coercion-missing-tail-expected-type.stderr]: I am not entirely sure as well. @nikomatsakis Should the label be reversed too?

File: src/librustc_typeck/check/coercion.rs (hunk 1 above)
Comment (nikomatsakis) [inline src/librustc_typeck/check/coercion.rs]: I think you can leave the comment, or amend some text like "(Note not all force-units work this way.)"

File: src/test/ui/coercion-missing-tail-expected-type.stderr (hunk 2 above)
Comment (nikomatsakis) [inline src/test/ui/coercion-missing-tail-expected-type.stderr]: Good catch @arielb1. To fix that you have to change this line (part of `fn coerce_inner()`, not found in the diff):

```rust
fcx.eq_types(true, cause, expression_ty, self.merged_ty())
```

to be something like this:

```rust
fcx.eq_types(label_expression_as_expected, cause, expression_ty, self.merged_ty())
```

That `true` argument indicates that the `a` value (first type) ought to be marked as "expected". That is no longer *always* true.

Comment (alexeyzab) [inline src/test/ui/coercion-missing-tail-expected-type.stderr]: Great, thanks for pointing this out!

**Pull request merged**